wcfsetup/install/files/ts/WoltLabSuite/Core/Media/Manager/Editor.ts

```diff
@@ -0,0 +1,395 @@
+      const insertIcon = listItem.querySelector(".jsMediaInsertButton");
+      if (insertIcon) {
+        insertIcon.classList.remove("jsMediaInsertButton");
+        insertIcon.addEventListener("click", (ev) => this._openInsertDialog(ev));
+      }
+    });
+  }
+
+  /**
+   * Builds the dialog to setup inserting media files.
+   */
+  protected _buildInsertDialog(): void {
+    let thumbnailOptions = "";
+
+    this._getThumbnailSizes().forEach((thumbnailSize) => {
+      thumbnailOptions +=
+        '<option value="' +
+        thumbnailSize +
+        '">' +
+        Language.get("wcf.media.insert.imageSize." + thumbnailSize) +
+        "</option>";
+    });
+    thumbnailOptions += '<option value="original">' + Language.get("wcf.media.insert.imageSize.original") + "</option>";
+
+    const dialog = `
+      <div class="section">
+        <dl class="thumbnailSizeSelection">
+          <dt>${Language.get("wcf.media.insert.imageSize")}</dt>
+          <dd>
+            <select name="thumbnailSize">
+              ${thumbnailOptions}
+            </select>
+          </dd>
+        </dl>
+      </div>
+      <div class="formSubmit">
+        <button class="buttonPrimary">${Language.get("wcf.global.button.insert")}</button>
+      </div>`;
+
+    UiDialog.open({
+      _dialogSetup: () => {
+        return {
+          id: this._getInsertDialogId(),
+          options: {
+            onClose: () => this._editorClose(),
+            onSetup: (content) => {
+              content.querySelector(".buttonPrimary")!.addEventListener("click", (ev) => this._insertMedia(ev));
+
+              DomUtil.show(content.querySelector(".thumbnailSizeSelection") as HTMLElement);
+            },
+            title: Language.get("wcf.media.insert"),
+          },
+          source: dialog,
+        };
+      },
+    });
+  }
+
+  protected _click(event: Event): void {
+    this._activeButton = event.currentTarget;
+
+    super._click(event);
+  }
+
+  protected _dialogShow(): void {
+    super._dialogShow();
+
+    // check if data needs to be uploaded
+    if (this._uploadData) {
+      const fileUploadData = this._uploadData as OnDropPayload;
+      if (fileUploadData.file) {
+        this._upload.uploadFile(fileUploadData.file);
+      } else {
+        const blobUploadData = this._uploadData as PasteFromClipboard;
+        this._uploadId = this._upload.uploadBlob(blobUploadData.blob);
+      }
+
+      this._uploadData = null;
+    }
+  }
+
+  /**
+   * Handles pasting and dragging and dropping files into the editor.
+   */
+  protected _editorUpload(data: OnDropPayload): void {
+    this._uploadData = data;
+
+    UiDialog.open(this);
+  }
+
+  /**
+   * Returns the id of the insert dialog based on the media files to be inserted.
+   */
+  protected _getInsertDialogId(): string {
+    let dialogId = "mediaInsert";
+
+    this._mediaToInsert.forEach((media, mediaId) => {
+      dialogId += `-${mediaId}`;
+    });
+
+    return dialogId;
```

Comment:
It probably doesn't really matter here, but repeated string concatenation can be slow. Building an `Array` and then joining the elements is preferred:

```
["mediaInsert", ...this._mediaToInsert.keys()].join("-")
```